### ISSUE TITLE
PolymorphicAuthDynamicFeature Usage/Improvement

### ISSUE THREAD

**Issue by psamsotha** [dropwizard/dropwizard#1632]
[`PolymorphicAuthDynamicFeature`][2] was added in an earlier 1.0 RC, and it seems to cover use cases in [#1579][1], but I'm looking at the code, and there seems to be a flaw in the feature. In the Javadoc it states

>Note that this feature only pertains to _authentication_. To enable authorization, use `AuthDynamicFeature`.

This sounds like the author would like us to use the `PolymorphicAuthDynamicFeature` in concert _with_ `AuthDynamicFeature`. This doesn't really make much sense. The `AuthDynamicFeature` registers a filter that handles both authentication _and_ authorization. It is not used _only_ for authorization. Using both features would end up registering the filters twice, and having them called twice. 

And also how are we expected to register `AuthDynamicFeature`. Say we have two filters for polymorphic feature

```           
final BasicCredentialAuthFilter filterOne = new BasicCredentialAuthFilter.Builder<PrincipalOne>()
        .setAuthenticator(authenticatorOne)
        .buildAuthFilter();
            
final BasicCredentialAuthFilter filterTwo = new BasicCredentialAuthFilter.Builder<PrincipalTwo>()
        .setAuthenticator(authenticatorTwo)
        .buildAuthFilter();
            
register(new PolymorphicAuthDynamicFeature<>(ImmutableMap.of(
         PrincipalOne.class, filterOne,
         PrincipalTwo.class, filterTwo
)));
```

What do we do now with the `AuthDynamicFeature`? It only takes one filter. Are we supposed to register _two_ `AuthDynamicFeature`s? There's isn't any documentation on intended usage. 

@evnm It looks like you were the author of this. Did you have an intended usage for this being used along side authorization, or did your javadoc comment mean that if you need authorization don't use this, instead use the other auth feature?

With the current implementation of the `PolymorphicAuthDynamicFeature`, was there something stopping you from simply using the same registration algorithm used in the `AuthDymamicFeature`, where the filter registration handle both authentication and authorization.

Without testing anything out, it seems to me that we _can_ use the poly feature for both authentication and authorization. We just need to use the same registration algo used in the `AuthDynamicFeature`. I'll check out the possibility later, but I'd like to hear what @evnm has to say. Maybe I am missing something.

[1]: https://github.com/dropwizard/dropwizard/issues/1579

[2]: https://github.com/dropwizard/dropwizard/blob/v1.0.0-rc4/dropwizard-auth/src/main/java/io/dropwizard/auth/PolymorphicAuthDynamicFeature.java

**Comment by evnm** [comment]
Thanks for bringing this up, @psamsotha.

My primary concern when putting a469b20 together was to minimize the amount of complexity added to the dropwizard-auth package while adding support for my immediate use case (multi-`Principal` authentication). dropwizard-auth was _already_ a major source of confusion within Dropwizard, as evidenced by the high proportion of issues and mailing list threads created by users unsure how to use it. I didn't want to make it any worse, so opted to restrict my change to authentication and included a disclaimer stating as such.

This restraint resulted in an obvious deficiency, in terms of rote functionality. Your suggestion points in the direction of a fine solution, but I think that supporting both authentication and authorization for polymorphic `Principal`s could introduce great risk of misconfiguration. In such cases, use of the `RolesAllowed` annotation would require consumers to be diligent in exhaustively configuring their `AuthFilter`s with `Authorizer`s that enforce the desired role policy. Omission of a `AuthFilterBuilder.setAuthorizer()` call for a given `AuthFilter` would result in that filter's `Principal` type being handled by the default `PermitAllAuthorizer`, possibly resulting in endpoints being erroneously exposed. In this way, the feature would "fail open" in an undesirable way.

This could be avoided if we could enforce with a compile error the specification of both an authenticator _and_ an authorizer for all `AuthFilters` passed to `PolymorphicAuthDynamicFeature`.

**Comment by psamsotha** [comment]
Let's take a step back and take a look at how all these components interact. Maybe this will shed some light on how to proceed.

So the components involved with the `AuthDynamicFeature` don't actually _do_ any authorization. What actually does the authorization, is Jersey's internal [`RolesAllowedRequestFilter`][1] that gets registered in the `RolesAllowedDynamicFeature`. If you look at the code for the `RolesAllowedRequestFilter`, you'll see that all it does is get the `SecurityContext` and call its `isUserInRole` method.

The `SecurityContext` gets set up in the [`AuthFilter`][2], which gets invoked just before the `RolesAllowedRequestFilter`. After the authentication, the `AuthFilter` creates a `SecurityContext` with the `Authorizer`. The `SecurityContext#isUserInRole` simply delegates to the `Authorizer#authorize` method

```java
requestContext.setSecurityContext(new SecurityContext() {
    @Override
    public boolean isUserInRole(String role) {
        return authorizer.authorize(principal.get(), role);
    }
});
```

So with any `AuthFilter`, no matter how it's registered, whether directly through Jersey

```java
final BasicCredentialAuthFilter filter
        = new BasicCredentialAuthFilter.Builder<PrincipalOne>()
                .setAuthenticator(authenticatorOne)
                .setAuthorizer(authorizer)
                .buildAuthFilter();
env.jersey().register(filter);
env.jersey().register(RolesAllowedDynamicFeature.class);
```

Or through the `AuthDynamicFeature`

```java
final BasicCredentialAuthFilter filter = ...
env.jersey().register(new AuthDynamicFeature(filter));
env.jersey().register(RolesAllowedDynamicFeature.class);
```

the `AuthFilter` will set up the `SecurityContext` properly for the `RolesAllowedRequestFilter`.

The problem with registering the filter directly through Jersey is that it gets invoked for _all_ resources endpoints. That is why we use the `AuthDynamicFeature`; so that the feature can determine _which_ endpoints it should be registered with. That's really the only role of the `AuthDynamicFeature`.

The `RolesAllowedDynamicFeature` does something similar. It has its registration algorithm to determine which endpoints should get the `RolesAllowedRequestFilter` registered, based on the authorization annotations. Because of this, if the filter is registered with an endpoint for which we do not have an `AuthFilter` also registered, then the authorization will fail, as the `SecurityContext` never gets set up.

All that being said, without changing _anything_, it is still possible to use the `PolymorphicAuthDynamicFeature` and still have authorization. The `PolymorphicAuthDynamicFeature` will register the authentication filter for each endpoint with an `@Auth` principal parameter

```java
@GET
public Response get(@Auth SomePrincipal principal) {}
```

If we want authorization on the same endpoint, we can just add a `@RolesAllowed` annotation, and the `RolesAllowedDynamicFeature` will automatically register the authz filter for this endpoint.

```java
@GET
@RolesAllowed({ "ADMIN "})
public Response get(@Auth SomePrincipal principal) {}
```

@evnm Let's try and break down you points of worry.

>In such cases, use of the `RolesAllowed` annotation would require consumers to be diligent in exhaustively configuring their `AuthFilters` with `Authorizers` that enforce the desired role policy. Omission of a `AuthFilterBuilder.setAuthorizer()` call for a given `AuthFilter` would result in that filter's Principal type being handled by the default `PermitAllAuthorizer`

As mentioned previously the `XAuthDynamicFeature`'s role is simply to determine which endpoints should get registered an _authentication_ filter. The `RolesAllowedDynamicFeature`, is what actually registers the authorization filter based on the Roles annotations.. So in the current state, your same worry applies to just using the `AuthFilter`, with or without an `XAuthDynamicFeature`. If the user register the `RolesAllowedDynamicFeature` and doesn't register an `Authorizer`, then yes, it will use the default `PermitAllAuthorizer`. 

```java
final BasicCredentialAuthFilter filter
        = new BasicCredentialAuthFilter.Builder<PrincipalOne>()
                .setAuthenticator(authenticatorOne)
                //.setAuthorizer(authorizer)  // no authorizer
                .buildAuthFilter();
env.jersey().register(filter);
env.jersey().register(RolesAllowedDynamicFeature.class);
```

Here is the problem you are worried about. It applies to all cases. It doesn't matter what filter is registered or in what way. If the user fails to register an `Authorizer`, and they register the `RolesAllowedDynamicFeature`, then the `RolesAllowedRequestFilter` will ultimately call the `PermitAllAuthorizer`. It is up the user to understand this about the architecture in general, not just in regards to using an `XAuthDynamicFeature`.

Let's look at the point I was making about using the registration algorithm that's used in the `AuthDynamicFeature`, also for the `PolymorphicAuthDynamicFeature`. But let's first examine the current implementation of the `AuthDynamicFeature` and see how it determines which endpoint should get registered the authentication filter

```java
@Override
public void configure(ResourceInfo resourceInfo, FeatureContext context) {
	final AnnotatedMethod am = new AnnotatedMethod(resourceInfo.getResourceMethod());
	final Annotation[][] parameterAnnotations = am.getParameterAnnotations();
	//@DenyAll shouldn't be attached to classes
	final boolean annotationOnClass = (resourceInfo.getResourceClass().getAnnotation(RolesAllowed.class) != null) ||
		(resourceInfo.getResourceClass().getAnnotation(PermitAll.class) != null);
	final boolean annotationOnMethod = am.isAnnotationPresent(RolesAllowed.class) || am.isAnnotationPresent(DenyAll.class) ||
		am.isAnnotationPresent(PermitAll.class);

	if (annotationOnClass || annotationOnMethod) {
		context.register(authFilter);
	} else {
		for (Annotation[] annotations : parameterAnnotations) {
			for (Annotation annotation : annotations) {
				if (annotation instanceof Auth) {
					context.register(authFilter);
					return;
				}
			}
		}
	}
}
```

So we can see there are two types of checks

1. Check if the endpoint or it's class has any _authorization_ annotations.
2. Check if the resource method has an `@Auth` parameter.

These checks are mutually exclusive, with the authorization annotations taking precedence. This kind of makes sense because when we are using the `RolesAllowedDynamicFeature`, it will register its authorization filter for all resource/methods annotated with the authorization annotations. So we should make sure to register authentication, attempting to use the same registration algorithm as the `RolesAllowedDynamicFeature`. This way there is always authentication to set up the `SecurityContext` for the Jersey filter to call.

If there are no authorization annotations, then just having an `@Auth` parameter is enough to register the authentication filter. In this case, whether or not we use the `RolesAllowedDynamicFeature` doesn't matter, as the `RolesAllowedDynamicFeature` on registers authorization when authorization annotations are present.

But the `@Auth` annotation isn't the only we to enable authentication. We can simply annotate classes/methods with `@PermitAll`. This also kind of makes sense. Just because we have authentication doesn't always mean we need access to the principal. So with `@PermitAll`, whether or not we have an `@Auth` parameter, there is authentication.

Now with all that said, how does this effect the `PolymorphicAuthDynamicFeature`?

Well for one, I don't think it's feasible to allow the case of just using `@PermitAll` by itself. There's no way to determine which filter should get registered. Actually this is the case for all authorization annotations.

So that brings us to the point of how do we sync the registration of the authentication filter with the registration of the authorization filter by the `RolesAllowedDynamicFeature`, just like the `AuthDynamicFeature` does?

Well now that I think about it, I don't see how it's possible. The `AuthDynamicFeature` doesn't take the `@Auth` annotation into consideration, when registering by looking for authorization annotations. Remember I said they are mutually exclusive in the current implementation. With the poly feature we would need to somehow combine these two checks. And I'm not sure it really makes much sense to do so, or may add way too much complexity.

I guess sometimes you need to write your thoughts down to disprove your own theories :-)

So I guess that still leave us with the main problem about how do we get authorization with the poly feature? As mentioned above, just adding the authorization annotations on the resource method, along with using the `RolesAllowedDynamicFeature` is enough.

I guess for now we should just document how it could/should be done. I propose maybe something along the lines of

>To use authorization with the `PolymorphicAuthDynamicFeature`, you need to register the `RolesAllowedDynamicFeature` with Jersey/DropWizard. Unlike when using the `AuthDynamicFeature`, where authorization annotations are allowed on classes, only methods (with corresponding `@Auth` principal parameters) should be annotated with `@RolesAllowed`

>    
```java
@GET 
@RolesAllowed({ "ADMIN" })
public Response get(@Auth MyPrincipal principal) {}
```

>The `PolymorphicAuthDynamicFeature` MUST NOT be used with the `AuthDynamicFeature`. Only one should be used. Using both will cause undesired behavior.

Along with the basic usage of the `PolymorphicAuthDynamicFeature` in the documentation, I think the above should give users enough information to get started.

All of this is up for discussion. All the above provided information is to help others get a better picture of what's going on. If we're in agreement that the current implementation should just be left alone, and just documented, I guess all that's left is to document the poly feature.

I've provided a complete test case in [this gist][3]. Nothing is modified. Simply adding the authorization annotation along with the `RolesAllowedDynamicFeature` is enough to get it working correctly.

[1]: https://github.com/jersey/jersey/blob/2.23.1/core-server/src/main/java/org/glassfish/jersey/server/filter/RolesAllowedDynamicFeature.java#L120

[2]: https://github.com/dropwizard/dropwizard/blob/v1.0.0-rc4/dropwizard-auth/src/main/java/io/dropwizard/auth/AuthFilter.java#L144

[3]: https://gist.github.com/psamsotha/308da89a56cdd88fa8f0169545a3e995

**Comment by evnm** [comment]
Thanks for taking the time to write your thought process in such detail. I think we're in agreement as to the behavior of the components in question.

I'm in favor of documenting the subtleties of the existing implementation rather than trying to augment it to cover additional use cases. Especially without any open issues indicating taht additional functionality is required by users.

Since you've put in so much effort already, would you like to put a change together adding to the Javadoc of `PolymorphicAuthDynamicFeature`? I'd be happy to review a PR if you'd like to move the discussion to one.

**Comment by psamsotha** [comment]
@evnm I ended up just adding a section in the main documentation. Seems more fitting there. Also removed the sentence in the javadoc about using the `AuthDynamicFeature`

**Comment by psamsotha** [comment]
Resolved with #1637
